internal/controllers/machineset/machineset_controller_status.go

```diff
@@ -0,0 +1,314 @@
+// This implies that the code in this function should account for several edge cases e.g. machine being partially provisioned,
+// machine being partially deleted but also for running machines being disrupted e.g. by deleting the node.
+// Additionally, this func should ensure that the conditions managed by this controller are always set in order to
+// comply with the recommendation in the Kubernetes API guidelines.
+// Note: v1beta1 conditions are not managed by this func.
+func (r *Reconciler) reconcileV1Beta2Status(_ context.Context, s *scope) {
+	// Update the following fields in status from the machines list.
+	// - replicas
+	// - v1beta2.readyReplicas
+	// - v1beta2.availableReplicas
+	// Also records the names of machines which are in deletion for more than 30 minutes in s.staleDeletingMachines.
+	setReplicas(s.machineSet, s.machines)
+
+	// Conditions
+
+	// Update the ScalingUp and ScalingDown condition, require the above setReplicas function.
+	setScalingUpCondition(s.machineSet, s.machines, s.bootstrapObjectNotFound, s.infrastructureObjectNotFound)
+	setScalingDownCondition(s.machineSet, s.machines, s.bootstrapObjectNotFound, s.infrastructureObjectNotFound, s.owningMachineDeployment)
+
+	// MachinesReady condition: aggregate the Machine's Ready condition
+	setMachinesReadyCondition(s.machineSet, s.machines)
+
+	// MachinesUpToDate condition: aggregate the Machine's UpToDate condition
+	setMachinesUpToDateCondition(s.machineSet, s.machines)
+
+	// TODO Deleting
+}
+
+func setReplicas(ms *clusterv1.MachineSet, machines []*clusterv1.Machine) {
+	// Return early when machines is nil. It's not possible to calculate replica counters.
+	// Conditions may surface to be unknown in this case.
+	if machines == nil {
+		return
+	}
+
+	var readyReplicas, availableReplicas int32
+	for _, machine := range machines {
+		if v1beta2conditions.IsTrue(machine, clusterv1.MachineReadyV1Beta2Condition) {
+			readyReplicas++
+		}
+		if v1beta2conditions.IsTrue(machine, clusterv1.MachineAvailableV1Beta2Condition) {
+			availableReplicas++
+		}
+	}
+
+	if ms.Status.V1Beta2 == nil {
+		ms.Status.V1Beta2 = &clusterv1.MachineSetV1Beta2Status{}
+	}
+
+	ms.Status.V1Beta2.ReadyReplicas = ptr.To(readyReplicas)
+	ms.Status.V1Beta2.AvailableReplicas = ptr.To(availableReplicas)
+}
+
+func setScalingUpCondition(ms *clusterv1.MachineSet, machines []*clusterv1.Machine, bootstrapObjectNotFound, infrastructureObjectNotFound bool) {
+	// If we got unexpected errors in listing the machines (this should happen rarely), surface them
+	if machines == nil {
+		v1beta2conditions.Set(ms, metav1.Condition{
+			Type:    clusterv1.MachineSetScalingUpV1Beta2Condition,
+			Status:  metav1.ConditionUnknown,
+			Reason:  clusterv1.MachineSetScalingUpInternalErrorV1Beta2Reason,
+			Message: "Please check controller logs for errors",
+		})
+		return
+	}
+
+	// Surface if .spec.replicas is not yet set (this should happen rarely).
```

Comment:
```suggestion
	// Surface if .spec.replicas is not yet set (this should never happen).
```
(same in other places if we have more of them)

I would drop the next sentence. Now after I took a closer look I don't think this happens with autoscaling. This should only ever happen if someone removes the MS webhook

Reply:
No strong opinion, but I would be also fine just combining this with the case above as an internal error (given that this case will never actually happen in reality)